Adding options to clear bucket of all files

skeetio opened the pull request:
Added an option to clear a bucket of all files.  Added a user prompt that they must verify before it continues and an optional argument to disable this prompt.

ppolewicz commented on b2:
```diff
@@ -1499,6 +1505,43 @@ class ConsoleTool(object):
+            elif user_input in ['N', 'NO']:
+                return
+
+        bucket = self.api.get_bucket_by_name(bucket_name)
```
This should be moved to the Bucket class

skeetio commented:
As you wish.  Moved the clear bucket logic into the Bucket class and added some user feedback using tqdm if available or just a simple print of the file name as it is being deleted.

ppolewicz commented on b2:
```diff
@@ -641,6 +647,36 @@ class Bucket(object):
+        while True:
+            response = self.list_file_names(start_filename=last_file, max_entries=100)
+
+            if len(response['files']) == 0:
```
```py
if not response['files']:
```

ppolewicz commented:
@skeetio thank you for your effort, it is really appreciated.

I noticed that after the progress bar change, API is using tqdm or print. The API itself is designed to be used in other programs, like FUSE filesystem for example, or a GUI tool, web page or something.

In the past every method handled errors with `sys.exit(1)`, which would terminate any of the above.

For this reason, we isolated the user interface from the actual logic, so that `ConsoleTool` can be used in case the tool is invoked from the shell and in other case the API and Bucket classes provide all input and output other interfaces could use.

Now, after the last change, this separation has been compromised, which would result in unwanted behavior in any application (other than `ConsoleTool`).

Would you be willing to work with me to improve it further, so that the progress bar functionality of `Bucket.clear()` and isolation of user interface are both preserved?

skeetio commented:
Would it be acceptable to implement this similar to upload_file with an optional parameter that defaults to suppressing printed/tqdm output?

ppolewicz commented:
"Acceptable" is a very good word to describe this solution. Ideally we would pass a progress reporter object, which would default to `None`, which would then automatically create a dummy object which doesn't do anything, and use this one. In case of `ConsoleTool`, a smarter progress reporter (based on `tqdm` or simple prints) would be used.
Then, in case of GUI application or web interface, some other progress reporter implementation could be plugged to report the progress. Current `upload_file` implementation is a kind of a hack that I hope we will eradicate (using the approach described above), so it would be best not to clone this approach further. If you don't have resources to implement it in a fully flexible way, the `quiet` flag will have to do for now and we will change it later.

skeetio commented:
Let me see what I can do.  Would it be your preference that the progress object be passed at the class level to be used by any methods as they see fit or to individual methods on an as-needed basis?

ppolewicz commented:
We must ask ourselves a question. If we were using this code as a library in our program, would we rather pass the progress reporter object to the `B2Api` constructor, or would we rather want to pass it to the methods? (this makes me lean toward passing it once)

What about concurrency? If we would pass a single progress reported object to `B2Api`, it would receive progress reports in an interleaved manner for each thread/greenlet running in parallel. (this makes me lean toward passing it in each method).

Maybe for the case of threading, we should pass a progress reporter factory into B2Api, which would then create a progress reporter for each file? That becomes complicated. I don't yet have a good solution in mind for this.

bwbeach commented:
I'm glad you're thinking about concurrency.  

I think I would favor passing the progress reporter to the B2Api constructor, assuming that we define a thread-safe progress reporting interface.  Maybe the B2Api constructor should get a progress reporter factory, so that it can use a separate progress reporter object for each action.

ppolewicz commented:
@skeetio We seem to have a similar idea with Brian about progress reporter factory (passed to B2Api constructor). It would be best approach considering future features.

bwbeach commented:
On the other hand, passing a progress reporter to each method in B2Api would let the application code pick the title string for each reporter.  If B2Api is calling a factory, it will either have to provide the title string for the progress reporter, or tell the factory what it's about to do (e.g., upload a file).  Passing a progress reporter into the upload_file method would keep B2Api simpler.  I think I've changed my mind, and now prefer this approach.

ppolewicz commented:
Now that you mention it, I noticed that passing state between progress reporter factory and api caller would be difficult. It would be much easier to pass it explicitly to upload_file etc

skeetio commented:
I feel like the scope of this has progressed far beyond what the original intent of the PR was, which was to add a mechanism to clear out a bucket, which is a prerequisite for bucket deletion. From the perspective of a user who had buckets in excess of 500+ files there was no good way to completely wipe and delete a bucket (including the web ui, which had timeout issues with deletes as small as 30 files).

How about we strip out the progress reporting aspect of the clear bucket method and tackle a more robust reporter implementation in a separate pr/branch?

ppolewicz commented:
I approve. Tbh I was surprised to see progress reporting in the same PR as clear_bucket.

ppolewicz commented:
By the way, would you consider adding a test for clear_bucket?

skeetio commented:
Will do. Also will add the method to the bash completions

On Sat, Feb 13, 2016, 2:39 PM Paweł Polewicz <notifications@github.com>
wrote:

> By the way, would you consider adding a test for clear_bucket?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Backblaze/B2_Command_Line_Tool/pull/58#issuecomment-183733044>
> .
>

ppolewicz commented:
Hi @skeetio,

I noticed that some time has passed in this issue. Do you intend to finish this?

bwbeach commented:
No action for a long time.  Closing.

nilayp commented:
Just wanted to know there is a "workaround" to clearing a bucket of all it's files using B2 CLI's sync command. It is documented on the Backblaze B2 knowledge base: https://help.backblaze.com/hc/en-us/articles/225556127-How-Can-I-Easily-Delete-All-Files-in-a-Bucket-
